foxai/explainer/computer_vision/algorithm/xrai.py

```diff
@@ -0,0 +1,692 @@
+"""File with XRAI algorithm explainer classes.
+
+Paper: https://arxiv.org/abs/1906.02825
+Based on https://github.com/PAIR-code/saliency/blob/master/saliency/core/xrai.py.
+"""
+from __future__ import absolute_import, division, print_function
+
+from typing import Any, Callable, Iterable, List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+from captum._utils.typing import TargetType
+from skimage import segmentation
+from skimage.morphology import dilation, disk
+from skimage.transform import resize
+
+from foxai.array_utils import validate_result
+from foxai.explainer.base_explainer import Explainer
+from foxai.explainer.computer_vision.algorithm.integrated_gradients import (
+    IntegratedGradientsCVExplainer,
+)
+from foxai.types import AttributionsType, ModelType
+
+
+def _normalize_image(
+    image_batch: np.ndarray,
+    value_range: Tuple[float, float],
+    resize_shape: Optional[Tuple[int, int]] = None,
+) -> np.ndarray:
+    """Normalize an image by resizing and rescaling its values.
+
+    Args:
+        image: Input images batch of shape `B x H x W x C`.
+        value_range: Value range to rescale values in form: `[min_value, max_value]`.
+        resize_shape: New images shape. If `None` images won't be resized.
+            Defaults to `None`.
+
+    Returns:
+        Resized and rescaled images batch.
+    """
+    image_max = np.max(image_batch)
+    image_min = np.min(image_batch)
+    image_batch = (image_batch - image_min) / (image_max - image_min)
+    image_batch = image_batch * (value_range[1] - value_range[0]) + value_range[0]
+    color_channels: int = image_batch.shape[-1]
+    resized_image_list: List[np.ndarray] = []
+    for image in image_batch:
+        if resize_shape is not None:
+            # resize can only process single image with shape H x W x C
+            resized_image_list.append(
```

Comment:
Do I understand correctly that resized_image_list will be empty by default? Meaning the result also will be empty by default (default resize_shape is None). Why is resize_shape Optional then if when not provided, the whole method doesn't make sense?

Reply:
Fixed by splitting this function into two: https://github.com/softwaremill/FoXAI/pull/99/commits/d13394acecd5f292cb4c1d59eb03f18d16b93b4c.

Reply:
I really like this change, it is much better.